zapcore/write_syncer_impl_test.go

```diff
@@ -0,0 +1,191 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zapcore
+
+import (
+	"bytes"
+	"errors"
+	"testing"
+
+	"io"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/testutils"
+)
+
+type writeSyncSpy struct {
+	io.Writer
+	testutils.Syncer
+}
+
+func requireWriteWorks(t testing.TB, ws WriteSyncer) {
+	n, err := ws.Write([]byte("foo"))
+	require.NoError(t, err, "Unexpected error writing to WriteSyncer.")
+	require.Equal(t, 3, n, "Wrote an unexpected number of bytes.")
+}
+
+func TestAddSyncWriteSyncer(t *testing.T) {
+	buf := &bytes.Buffer{}
+	concrete := &writeSyncSpy{Writer: buf}
+	ws := AddSync(concrete)
+	requireWriteWorks(t, ws)
+
+	require.NoError(t, ws.Sync(), "Unexpected error syncing a WriteSyncer.")
+	require.True(t, concrete.Called(), "Expected to dispatch to concrete type's Sync method.")
+
+	concrete.SetError(errors.New("fail"))
+	assert.Error(t, ws.Sync(), "Expected to propagate errors from concrete type's Sync method.")
+}
+
+func TestAddSyncWriter(t *testing.T) {
+	// If we pass a plain io.Writer, make sure that we still get a WriteSyncer
+	// with a no-op Sync.
+	buf := &bytes.Buffer{}
+	ws := AddSync(buf)
+	requireWriteWorks(t, ws)
+	assert.NoError(t, ws.Sync(), "Unexpected error calling a no-op Sync method.")
+}
+
+func TestBufferedWriterWrite(t *testing.T) {
+	tests := []struct {
+		desc    string
+		wrap    func(WriteSyncer) WriteSyncer // wrap or replace the underlying WriteSyncer
+		initial string                        // initial buffer contents
+		try     string                        // bytes to write
+		buf     string                        // final buffer contents
+		flushed string                        // written to underlying WriteSyncer
+		n       int
+		err     error
+	}{
+		{
+			desc:    "new write fills buffer",
+			initial: "ab",
+			try:     "cd",
+			buf:     "abcd",
+			flushed: "",
+			n:       2,
+		},
+		{
+			desc:    "new write triggers buffer flush",
+			initial: "ab",
+			try:     "cde",
+			buf:     "cde",
+			flushed: "ab",
+			n:       3,
+		},
+		{
+			desc:    "new write is too big to buffer",
+			initial: "ab",
+			try:     "cdefg",
```

Comment:
have another case where the try is just "abc" from this case, it should still be "too big" because of "remaining space" where-as a 5-length string would've never fit irrespective of initial